app/src/main/java/com/android/streetworkapp/model/image/ImageRepositoryFirestore.kt

```diff
@@ -112,4 +107,230 @@ class ImageRepositoryFirestore(
+   *
+   * @param imageCollectionId The collection id the image is part of.
+   * @param imageUrl The url of the image.
+   * @param userId The userId of the vote to remove.
+   */
+  override suspend fun retractImageVote(
+      imageCollectionId: String,
+      imageUrl: String,
+      userId: String
+  ): Boolean {
+    try {
+      val docRef =
+          db.collection(ImageRepositoryFirestore.COLLECTION_PATH).document(imageCollectionId)
+      val document = docRef.get().await()
+      val imageCollection = document.toObject(ParkImageCollection::class.java)
+
+      // pretty inefficient to go through the whole list but the data will be pretty small anyways
+      imageCollection?.images?.let {
+        val updatedImages =
+            it.map { image ->
+              if (image.imageUrl == imageUrl) { // identifying the images by their url
+                val updatedRating: ImageRating?
+                if (image.rating.positiveVotesUids.contains(userId)) {
+                  updatedRating =
+                      image.rating.copy(
+                          positiveVotes = image.rating.positiveVotes - 1,
+                          positiveVotesUids = image.rating.positiveVotesUids - userId)
+                } else if (image.rating.negativeVotesUids.contains(userId)) {
+                  updatedRating =
+                      image.rating.copy(
+                          negativeVotes = image.rating.negativeVotes - 1,
+                          negativeVotesUids = image.rating.positiveVotesUids - userId)
+                } else {
+                  return false // userId was not in the list, return false
+                }
+
+                return@map image.copy(rating = updatedRating)
+              } else {
+                return@map image // No change for this image
+              }
+            }
+
+        docRef.update("images", updatedImages).await()
+        return true
+      } ?: return false
+    } catch (e: Exception) {
+      Log.d(
+          DEBUG_PREFIX,
+          e.message
+              ?: "An exception occurred but the message associated with it couldn't be retrieved.")
+      return false
+    }
+  }
+
+  /**
+   * Deletes all the images related to a user
+   *
+   * @param userId The user id to whom we delete all the data from. (pictures uploaded and ratings)
+   */
+  override suspend fun deleteAllDataFromUser(userId: String): Boolean {
+    require(userId.isNotEmpty()) { "Empty userId." }
+
+    try {
+      val querySnapshot =
+          db.collection(ImageRepositoryFirestore.COLLECTION_PATH)
+              .get()
+              .await() // get all documents from collection
+
+      for (document in querySnapshot.documents) {
+        val collection = document.toObject(ParkImageCollection::class.java)
+        collection?.let {
+          for (image in collection.images) {
+            if (image.userId == userId) {
+              document.reference.update("images", FieldValue.arrayRemove(image)).await()
```

Comment:
I saw that you could do something like this

```kotlin

         val updatedImages = collection.images.filterNot { it.userId == userId }
         val docRef = document.reference
         // Update the images in Firestore
         batch.update(docRef, "images", updatedImages)
         // Delete the associated storage objects
         collection.images.filter { it.userId == userId }.forEach { image ->
             storageClient.extractKeyFromUrl(image.imageUrl)?.let { storageClient.deleteObjectFromKey(it) }
         }
```

Reply:
I've done something in the same spirit:
`val (imagesUploadedByUser, imagesNotUploadedByUser) =
              collection.images.partition { it.userId == userId }
          document.reference.update("images", FieldValue.arrayRemove(imagesUploadedByUser)`

And I then iterate `imagesNotUploadedByUser` to remove his votes.